waddrmgr/manager_test.go

```diff
@@ -3137,3 +3137,148 @@ func assertAddressDerivation(t *testing.T, db walletdb.DB,
 	})
 	require.NoError(t, err)
 }
+
+// TestManagedAddressValidation tests that we're able to properly detect and
+// reject various fault cases for address/key derivation.
+func TestManagedAddressValidation(t *testing.T) {
+	t.Parallel()
+
+	teardown, db := emptyDB(t)
+	defer teardown()
+
+	// We'll start the test by creating a new root manager that will be
+	// used for the duration of the test.
+	var mgr *Manager
+	err := walletdb.Update(db, func(tx walletdb.ReadWriteTx) error {
+		ns, err := tx.CreateTopLevelBucket(waddrmgrNamespaceKey)
+		if err != nil {
+			return err
+		}
+		err = Create(
+			ns, rootKey, pubPassphrase, privPassphrase,
+			&chaincfg.MainNetParams, fastScrypt, time.Time{},
+		)
+		if err != nil {
+			return err
+		}
+		mgr, err = Open(ns, pubPassphrase, &chaincfg.MainNetParams)
+		if err != nil {
+			return err
+		}
+
+		return mgr.Unlock(ns, privPassphrase)
+	})
+	require.NoError(t, err, "create/open: unexpected error: %v", err)
+
+	defer mgr.Close()
+
+	var testCases = []struct {
+		name        string
+		mutator     func(addr *managedAddress, ogPrivKey *btcec.PrivateKey)
+		expectedErr error
+	}{
+		{
+			// In this case, we'll swap out the public key with some other
+			// random public key.
+			name: "derive diff pubkey",
+			mutator: func(addr *managedAddress, priv *btcec.PrivateKey) {
+				newPriv, err := btcec.NewPrivateKey()
+				if err != nil {
+					t.Fatalf("unable to gen priv key: %v", err)
+				}
+
+				addr.pubKey = newPriv.PubKey()
+			},
+			expectedErr: ErrPubkeyMismatch,
+		},
+		{
+			// In this case, we'll swap out the contained address another
+			// address that's actually a different type. This simulates a
+			// bitflip during the address generation process.
+			name: "addr bitflip",
+			mutator: func(addr *managedAddress, priv *btcec.PrivateKey) {
+				hash160 := btcutil.Hash160(
+					addr.pubKey.SerializeCompressed(),
+				)
+				hash160[0] ^= 0x01
+
+				addr.address, err = btcutil.NewAddressWitnessPubKeyHash(
+					hash160, &chaincfg.MainNetParams,
+				)
+				if err != nil {
+					t.Fatalf("unable to gen addr: %v", err)
+				}
+			},
+			expectedErr: ErrAddrMismatch,
+		},
+		{
+			// In this final case, we'll have the signature generator
+			// return a signature that's actually invalid for the public
+			// key.
+			name: "wrong priv key",
+			mutator: func(addr *managedAddress, priv *btcec.PrivateKey) {
+				// To trigger an invalid private key, we'll
+				// modify the internal decrypted private key.
+				addr.privKeyCT[0] ^= 0x01
```

Comment:
Ah, I see why you used the `privKeyCT` directly before. But I think this test should still work the same when using `PrivKey()` in `Validate()`.

Reply:
See above re the issue w/ `PrivKey` and one of the tests. 

I also had things flipped: I'd use the passed in private key to generate a signature against the public key of the addr. However, that made it hard to write a tests here as modifying the passed in private key would actually cause the earlier tests to fail. 

One way to get around this would be to abstract out the calls to `Sign` behind some interface, then have that pass a bogus signature.